Tensile/KernelWriter.py

```diff
@@ -3422,31 +3469,33 @@ def kernelBody( self, kernel, tensorParametersA, tensorParametersB ):
               kl.append(self.comment("tail loop unroll iter %u"%(mValue)))
             pack[0] = Code.Module()
+            # always use vregSetIdx=0 for DirectToVgpr + tail loop
+            vregSetIdxMFMA = 0
             for iui in range(0, tailLoopInnerUnroll):
               if self.enable["LocalRead"]:
-                doReadA = not kernel["DirectToVgprA"]
-                doReadB = not kernel["DirectToVgprB"]
                 # local read buffer id. No prefetch in tail loop case.
                 bufIdx = (mValue % (self.numVgprBuffer+1)) % kernel["LoopIters"]
-                if mValue*self.numReadsIterCoalescedA < mEnd and doReadA:
+                # DTVpack case, use different bufIdx for all loop iterations
+                bufIdxDTV = mValue % kernel["LoopIters"]
+                bufIdxA = bufIdxDTV if self.packDTVA else bufIdx
+                bufIdxB = bufIdxDTV if self.packDTVB else bufIdx
+                if mValue*self.numReadsIterCoalescedA < mEnd:
                   # Reading 16-bit data from LDS requires packing when ECC enabled
                   kl.append(self.comment("local read a"))
-                  localReadCodeA, packCodeA = self.localReadDo(kernel, bufIdx*self.numIterPerCoalescedReadA, iui*self.numIterPerCoalescedReadA, 0, tensorParametersA)
+                  localReadCodeA, packCodeA = self.localReadDo(kernel, bufIdxA*self.numIterPerCoalescedReadA, iui*self.numIterPerCoalescedReadA, 0, tensorParametersA)
                   kl.append(localReadCodeA)
                   pack[0].addCode(packCodeA)
-                if mValue*self.numReadsIterCoalescedB < mEnd and doReadB:
+                if mValue*self.numReadsIterCoalescedB < mEnd:
                   kl.append(self.comment("local read b"))
-                  localReadCodeB, packCodeB = self.localReadDo(kernel, bufIdx*self.numIterPerCoalescedReadB, iui*self.numIterPerCoalescedReadB, 0, tensorParametersB)
+                  localReadCodeB, packCodeB = self.localReadDo(kernel, bufIdxB*self.numIterPerCoalescedReadB, iui*self.numIterPerCoalescedReadB, 0, tensorParametersB)
                   kl.append(localReadCodeB)
                   pack[0].addCode(packCodeB)
                 # adjustment for DirectToLds case
                 iuiParam = iui + tailLoopInnerUnroll * mValue
-                if doReadA:
-                  kl.append(self.comment("local read inc a"))
-                  kl.append(self.localReadInc(kernel, iuiParam, tensorParametersA))
-                if doReadB:
-                  kl.append(self.comment("local read inc b"))
-                  kl.append(self.localReadInc(kernel, iuiParam, tensorParametersB))
+                kl.append(self.comment("local read inc a"))
+                kl.append(self.localReadInc(kernel, iuiParam, tensorParametersA))
+                kl.append(self.comment("local read inc b"))
+                kl.append(self.localReadInc(kernel, iuiParam, tensorParametersB))
```

Comment:
This always generates dtva and dtvb code now?

Reply:
localReadInc does nothing in DirectToVgpr case.